### PR TITLE
Refactor, add new options, better error handling

### PR DESCRIPTION
**Pull request by joshuathompson**

### CODE REVIEW

File: lib/stream-to-mongo.js [hunk 1]
```diff
@@ -12,33 +13,35 @@ export default class StreamToMongo extends Writable {
     }
 
     _write(chunk, encoding, next) {
+        this.startWrite(chunk, next);
```
Comment (perryao) [inline lib/stream-to-mongo.js]: since startWrite is a promise, is this going to actually resolve?

File: lib/index.js [hunk 2]
```diff
@@ -1,45 +1,47 @@
...
+        const options = {
+            db: connectionString,
+            collection: config.collection,
+            uniqueAttribute: config.uniqueAttribute,
+            drop: config.drop,
+        };
+        const writeStream = new StreamToMongo(options);
         stream.pipe(JSONStream.parse()).pipe(writeStream);
-        return new Promise((resolve, reject) => {
+        return new Promise((resolve) => {
```
Comment (perryao) [inline lib/index.js]: should we not handle handle stream.on('error') and reject it? Also, I've ran into issues before resolving the promise on finish. The stream itself is listening to finish and closing the db, but if we resolve before that happens, might not fully flush and close the connection. In other write streams, I've changed to emitting a 'close' event when done writing data and using that event to resolve the promise.

File: lib/stream-to-mongo.js (hunk 1 above)
Comment (JoshuaThompson) [inline lib/stream-to-mongo.js]: `startWrite` isn't a promise, it's an async method.

Comment (perryao) [inline lib/stream-to-mongo.js]: the result of calling an async method is a Promise object.

Comment (perryao) [inline lib/stream-to-mongo.js]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function
More exactly, an AsyncFunction.

Comment (JoshuaThompson) [inline lib/stream-to-mongo.js]: Noted.  As we discussed it technically did work as it was however it's better practice to remove ambiguity and use promises more correctly without mixing with callbacks.

File: lib/stream-to-mongo.js [hunk 3]
```diff
@@ -12,49 +14,62 @@ export default class StreamToMongo extends Writable {
...
+            })
+            .catch((err) => {
+                this.log.error(err);
+                next(err);
             });
+    }
+
+    async startWrite(chunk) {
+        if (!this.db) {
+            await this.connectToDb().catch((err) => { this.log.info(err); });
```
Comment (perryao) [inline lib/stream-to-mongo.js]: since we're catching this error here, it won't ever be propagated back up in the catch block in startWrite

Comment (perryao) [inline lib/stream-to-mongo.js]: should probably just remove the catch here.

**Pull request merged**